File: composeApp/src/commonMain/kotlin/com/daddoodev/yetimatch/ui/screens/HomeScreen.kt
package com.daddoodev.yetimatch.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.daddoodev.yetimatch.models.Category
import com.daddoodev.yetimatch.models.QuizManifest
import com.daddoodev.yetimatch.models.QuizMeta
import com.daddoodev.yetimatch.ui.YetiMatchLogo

@Composable
fun CategoryCard(
    category: Category,
    onClick: () -> Unit
) {
    Card(
        onClick = onClick,
        modifier = Modifier
            .fillMaxWidth()
            .padding(4.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant),
        elevation = CardDefaults.cardElevation(defaultElevation = 2.dp)
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Text(
                text = category.name,
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.primary
            )
            Text(
                text = category.description,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

@Composable
fun HomeScreen(
    manifest: QuizManifest?,
    searchQuery: String,
    onSearchQueryChange: (String) -> Unit,
    onCategoryClick: (String) -> Unit,
    onQuizClick: (QuizMeta) -> Unit
) {
    val filteredQuizzes = remember(manifest, searchQuery) {
        if (manifest == null || searchQuery.isBlank()) emptyList()
        else {
            val q = searchQuery.lowercase()
            val categoryNames = manifest.categories.associate { it.id to it.name.lowercase() }
            manifest.quizzes.filter { meta ->
                meta.title.lowercase().contains(q) ||
                    meta.description.lowercase().contains(q) ||
                    categoryNames[meta.categoryId]?.contains(q) == true
            }
        }
    }

    Column(modifier = Modifier.fillMaxWidth()) {
        YetiMatchLogo(size = 80.dp, modifier = Modifier.padding(top = 16.dp))
        Spacer(modifier = Modifier.height(16.dp))
        OutlinedTextField(
            value = searchQuery,
            onValueChange = onSearchQueryChange,
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            placeholder = { Text("Search quizzes...") },
            singleLine = true
        )

        if (searchQuery.isNotBlank()) {
            if (filteredQuizzes.isEmpty()) {
                Text(
                    text = "No quizzes match \"$searchQuery\"",
                    style = MaterialTheme.typography.bodyLarge,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(16.dp)
                )
            } else {
                LazyColumn(
                    contentPadding = PaddingValues(16.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    items(filteredQuizzes, key = { it.id }) { meta ->
                        QuizMetaCard(meta = meta, onClick = { onQuizClick(meta) })
                    }
                }
            }
        } else if (manifest != null) {
            LazyVerticalGrid(
                columns = GridCells.Fixed(2),
                contentPadding = PaddingValues(16.dp),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                items(manifest.categories, key = { it.id }) { cat ->
                    CategoryCard(category = cat, onClick = { onCategoryClick(cat.id) })
                }
            }
        }
    }
}

File: lib/screens/home_screen.dart
import 'package:flutter/material.dart';

import '../models/manifest.dart';
import '../theme/app_theme.dart';

class HomeScreen extends StatelessWidget {
  const HomeScreen({
    super.key,
    required this.manifest,
    required this.searchQuery,
    required this.onSearchQueryChange,
    required this.onCategoryClick,
    required this.onQuizClick,
  });

  final QuizManifest? manifest;
  final String searchQuery;
  final ValueChanged<String> onSearchQueryChange;
  final ValueChanged<String> onCategoryClick;
  final ValueChanged<QuizMeta> onQuizClick;

  @override
  Widget build(BuildContext context) {
    final filtered = _filterQuizzes(manifest, searchQuery);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        Padding(
          padding: const EdgeInsets.all(16),
          child: TextField(
            onChanged: onSearchQueryChange,
            decoration: const InputDecoration(
              hintText: 'Search quizzes...',
              border: OutlineInputBorder(),
            ),
          ),
        ),
        Expanded(
          child: searchQuery.isNotEmpty
              ? (filtered.isEmpty
                  ? Padding(
                      padding: const EdgeInsets.all(16),
                      child: Text(
                        'No quizzes match "$searchQuery"',
                        style: Theme.of(context).textTheme.bodyLarge,
                      ),
                    )
                  : ListView.builder(
                      padding: const EdgeInsets.all(16),
                      itemCount: filtered.length,
                      itemBuilder: (_, i) => _QuizMetaCard(
                        meta: filtered[i],
                        onTap: () => onQuizClick(filtered[i]),
                      ),
                    ))
              : (manifest == null
                  ? const SizedBox.shrink()
                  : ListView.builder(
                      padding: const EdgeInsets.all(16),
                      itemCount: (manifest!.categories.length + 1) ~/ 2,
                      itemBuilder: (_, rowIndex) {
                        final i = rowIndex * 2;
                        final left = manifest!.categories[i];
                        final right = i + 1 < manifest!.categories.length
                            ? manifest!.categories[i + 1]
                            : null;
                        return Padding(
                          padding: const EdgeInsets.only(bottom: 12),
                          child: Row(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: [
                              Expanded(
                                child: _CategoryCard(
                                  category: left,
                                  onTap: () =>
                                      onCategoryClick(left.id),
                                ),
                              ),
                              const SizedBox(width: 12),
                              Expanded(
                                child: right != null
                                    ? _CategoryCard(
                                        category: right,
                                        onTap: () =>
                                            onCategoryClick(right.id),
                                      )
                                    : const SizedBox.shrink(),
                              ),
                            ],
                          ),
                        );
                      },
                    )),
        ),
      ],
    );
  }

  List<QuizMeta> _filterQuizzes(QuizManifest? m, String query) {
    if (m == null || query.isEmpty) return [];
    final q = query.toLowerCase();
    final categoryNames = {
      for (final c in m.categories) c.id: c.name.toLowerCase(),
    };
    return m.quizzes.where((meta) {
      return meta.title.toLowerCase().contains(q) ||
          meta.description.toLowerCase().contains(q) ||
          (categoryNames[meta.categoryId]?.contains(q) ?? false);
    }).toList();
  }
}

class _CategoryCard extends StatelessWidget {
  const _CategoryCard({
    required this.category,
    required this.onTap,
  });

  final Category category;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final cardColor = categoryCardColor(
      theme.colorScheme,
      theme.brightness,
    );
    return Card(
      elevation: 2,
      color: cardColor,
      child: InkWell(
        onTap: onTap,
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            mainAxisSize: MainAxisSize.min,
            children: [
              Text(
                category.name,
                style: theme.textTheme.titleMedium?.copyWith(
                      color: theme.colorScheme.primary,
                    ),
              ),
              const SizedBox(height: 4),
              Text(
                category.description,
                style: theme.textTheme.bodySmall,
                maxLines: 4,
                overflow: TextOverflow.ellipsis,
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _QuizMetaCard extends StatelessWidget {
  const _QuizMetaCard({required this.meta, required this.onTap});

  final QuizMeta meta;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return Card(
      elevation: 2,
      margin: const EdgeInsets.only(bottom: 8),
      child: ListTile(
        title: Text(
          meta.title,
          style: Theme.of(context).textTheme.titleMedium?.copyWith(
                color: Theme.of(context).colorScheme.primary,
              ),
        ),
        subtitle: Text(meta.description),
        onTap: onTap,
      ),
    );
  }
}
